lib/r10k/api.rb

```diff
@@ -0,0 +1,160 @@
+    # resolved to specific versions (or commit SHAs).
+    #
+    # If passed an already resolved environment, this function will have no effect.
+    #
+    # This function assumes that all relevant caches have already been updated.
+    #
+    # @param env_map [Hash] A hashmap representing a single environment's desired state.
+    # @return [Hash] A copy of env_map with any ambiguous module versions or ranges replaced with specific versions and a :resolved_at timestamp added.
+    def resolve_environment(env_map)
+    end
+
+    # Given an environment map, write the base environment and all Puppetfile declared modules to disk at the given path.
+    #
+    # @param env_map [Hash] A fully-resolved (see {#resolve_environment}) hashmap representing a single environment's new state.
+    # @param path [String] Path on disk into which the given environment should be deployed. The given path should already include the environment's name. (e.g. /puppet/environments/production not /puppet/environments) Path will be created if it does not already exist.
+    # @param opts [Hash] Additional options as defined.
+    # @return [true] Returns true on success, raises on failure.
+    # @raise [RuntimeError] Something bad happened!
+    def write_environment!(env_map, path, opts={})
+    end
+
+    # Given an environment map, write the base environment only (not any Puppetfile declared modules) to disk at the given path.
+    #
+    # @param env_map [Hash] A fully-resolved (see {#resolve_environment}) hashmap representing a single environment's new state.
+    # @param path [String] Path on disk into which the given environment should be deployed. The given path should already include the environment's name. (e.g. /puppet/environments/production not /puppet/environments) Path will be created if it does not already exist.
+    # @param opts [Hash] Additional options as defined.
+    # @return [true] Returns true on success, raises on failure.
+    # @raise [RuntimeError] Something bad happened!
+    def write_env_base!(env_map, path, opts={})
+    end
+
+    # Write the given module, using the version/commit declared in the given environment map, to disk at the given path.
+    #
+    # @param module_slug [String] Hyphen separated namespace and module name of the module to be written to disk. (E.g. "puppetlabs-apache")
+    # @param env_map [Hash] A fully-resolved (see {#resolve_environment}) hashmap representing a single environment's new state.
+    # @param path [String] Path on disk into which the given module should be deployed. The given path should already include the environment's name. (e.g. /puppet/environments/production not /puppet/environments) Path will be created if it does not already exist.
+    # @param opts [Hash] Additional options as defined.
+    # @return [true] Returns true on success, raises on failure.
+    # @raise [RuntimeError] Something bad happened!
+    def write_module!(module_slug, env_map, path, opts={})
```

Comment:
You're correct, at this level it should just be module_name.

It doesn't need to be a full env_map, but I feel like the API is easier to consume if this function handles destructuring what it needs out of the map instead of making the consuming code do that.